search - beam and campaign

kudlohlavec opened the pull request:
Quick work summary:
- we are using [laravel scout](https://laravel.com/docs/5.7/scout) for search purposes
- scout works on top of search drivers, however default one is payed so we have switched it to free [tntsearch-driver](https://github.com/teamtnt/laravel-scout-tntsearch-driver)
- it is very robust search engine that contains features such as fuzzy search, typeahead support etc.
- it requires creating of initial indexing files for models that will be searchable - for this purpose we have created artisan script that is being called from Docker/php/remp.sh install script. After creation of indexing files, no further action is being required, changes in models are being tracked automatically 
- search results are being displayed as-you-type thanks to [jquery typeahead plugin](https://github.com/running-coder/jquery-typeahead)
- we have adopted current search box design

Thinks to consider:
- currently no search is being performed in sso project, so it makes sense to remove search bar in this project
- many aspects of tnt search driver and typeahead functionality are customisable. Parameters such as minimal string length for search init (currently 2 chars), maximum results being shown (currently unlimited), delay of search request resending during typing (currently 300 ms) are few ones to mention for consideration

kudlohlavec commented:
> Apart from the comments in the code:
> 
> * Please change placeholder texts in the search boxes.

 Can you please specify what it should exactly be so we can avoid multiple change requests on such trivialities

> * How often should the reindexing happen? Or is it done automatically when new records are created?

Its being done automatically on every change. However we should consider queueing of reindexing to speed up responsiveness of the app. Its little bit harder for me to replicate your production load so these are only guesses from my side, maybe it will be reindexed in an instant. One way or another its only matter of one config line change.

> * What's the support of indexing for multi-server installations? How does reindexing work there? Can the index be stored in DB/Redis?

Can you please give some example of multi-server architecture? If storing of index files should be an issue, its storage path can be defined explicitly, so it can be directed to some kind of shared volume that is accesible from multiple servers. However, indexes are being generated per-project, so if by multi-server architecture you meant different server per project it should be totally ok, thats the reason why I asked for example of muli-server architecture to see if it will be problem or not. Indexes can be stored in files only.

> 
> The initial index generation is running for an hour at this point (with our production data) and still going on. That feels like a lot but if the multi-server is handled and the index is updated on the fly later on, we would be probably OK with that. I'll post an update when it's finished. If it's not OK, this solution wouldn't work and we'll have to try something different or simpler (mysql fulltext?)

There is also another indexing command that should be faster for large dataset scenarios, I will look into it if its usable for us, because it got some limitations as well...

rootpd commented:
> Can you please specify what it should exactly be? So we can avoid multiple change requests on such trivialities

It depends based on what's being searched. If the search is prepared for articles, authors and segments, it should say "Search for articles, authors and segments", etc.

> However we should consider queueing of reindexing to speed up responsiveness of the app.

Update of search index shouldn't be blocking.

>  Its little bit harder for me to replicate your production load so these are only guesses from my side

We have ~70K articles, ~3K authors and tens of segments in the production database. You can use `db:seed` artisan command to generate fake data. Run it multiple times to get more random data.

> Can you please give some example of multi-server architecture? If storing of index files should be an issue, its storage path can be defined explicitly, so it can be directed to some kind of shared volume that is accesible from multiple servers.

The project/application (e.g. Beam) is being server by multiple PHP servers in multiple datacenters. Having index on the disk is useless as it would always get updated only on the application server that handled the request - or server handling the background queue.

Also having shared network volume between the datacenters is something that we cannot proceed with - it would be issue in our infrastructure and I can imagine it would be an issue for other publishers too. If something is stored on the disk, it should be local to the server (e.g. framework temp files). If something needs to be shared, it should be in DB/Redis.

If this solution cannot do this, something else will unfortunately need to be used. And I wouldn't probably try to bend this further (by extracting search to separate service), it really should be as simple as it could possibly be - it's an internal search in an internal app used by tens of people.

> There is also another indexing command that should be faster for large dataset scenarios, I will look into it if its usable for us, because it got some limitations as well...

The initial indexation took 2 hours and generated two files:

```
-rw-r--r-- 1 rootpd   rootpd   25845760 apr  1 01:45 articles.index
-rw-r--r-- 1 rootpd   rootpd     315392 apr  1 01:45 authors.index

```

kudlohlavec commented:
> If this solution cannot do this, something else will unfortunately need to be used.

Tnt driver works only on top of the sqlite files so this will be not the way for multi-server scenarios. I have been testing few other scout driver alternatives and one very promising uses elasticsearch. I have succesfully implemented it with your current elasticsearch cluster, however it needs further testing. Is it ok with you if indexing will be distributed by elasticsearch?

rootpd commented:
This really feels like an overkill (literally bazooka for shooting down the fly). I still have a feeling that fulltext index would do as good job for what it is expected to do.. 

If I search "korona", I expect it to find "koronavirus" tag and the newest articles with this tag. I really don't feel that such thing requires external database. That's the most advanced scenario to support.

kudlohlavec commented:
> This really feels like an overkill (literally bazooka for shooting down the fly). I still have a feeling that fulltext index would do as good job for what it is expected to do..

I was also testing mysql driver for scout that was leveraging fulltext index but it was failing on indexing of relations so I dropped it.

I was preferring usage of scout because it is native tool from laravel and aligns seamlessly into whole ecosystem along with leveraging of drivers with robust and popular search features. I consider it to be solid, reliable and robust solution also towards possible rising complexity of requirements for search in the future. 
But I understand that you got much more better understanding of whole product and its roadmap. For future tasks it will be for the best to clarify all of the aspects such as prioritizations of simplicity over robustness or today mentioned multi-server scenario to avoid misunderstandings.

I can solely look for another fulltext index solutions for laravel without usage of scout or will make it myself if nothing sufficient will be available.

rootpd commented:
Sure, understandable.

If possible, always work with presumption that we want to have as few external dependencies as possible. We evaluate all of them (and their cost/benefit tradeoffs) to understand how they impact our installation and also all other REMP installations.

And also probably with presumption that you're maintainer of such installation yourself. You wouldn't probably be very eager to update if you saw that you need to add network storage into your infrastructure (or another Elastic cluster) because of the search. Mainly if it's kind of the data that the CMS can search in without it.

rootpd commented:
@kudlohlavec One more thing. The same search will be needed for Mailer. Trying to use framework-independent solution will save time during Mailer implementation. Don't try to integrate "nicely" with Laravel, aim for simplicity.

kudlohlavec commented:
@rootpd How independent do you want it to be? Because as you surely know, each framework got its own ORM. So do you want us to write raw queries? Or did you mean to think about reusable parts such as same response structure from backend, requests, frontend parts such as js, css and html? Because all of the latter is already being reused...

kudlohlavec commented:
What about the number of returned results? I can imagine that on 70k articles there can be also queries that will return a lot of data. I propose to limit it to some reasonable number, lets say 10 results and order it by id to get latest records? Please let me know if it makes sense or do you want to enable for user to see all of the results by some GUI element that will redirect to another result page? Or make result element scrollable?

rootpd commented on Beam/app/Http/Controllers/SearchController.php:
```diff
@@ -0,0 +1,32 @@
+        $stopwatch->start('search');
+        $searchResult['articles'] = Article::search($searchTerm)->with(['tags', 'sections'])->get();
+        $time = $stopwatch->stop('search');
+        dd($searchResult['articles'], $time->getDuration());
```
This one looks forgotten..

kudlohlavec replied:
its just for testing purposes, we got push mirroring on our repos between gitlab and github so once we create PR it also reflects all commits here... i have purposely stated WIP keyword in the latest commit because of that nature of commit 😄

rootpd replied:
oh, ok :). sorry :)

rootpd commented:
Regarding sorting and limiting, let's do limit. I'd say it's fine to return last 5 records (per each group if possible). Sorting is also a good idea, sorting field will depend on the type of returned element.

* article - published_at
* tag/section/author - based on the last published article
* email/layout/job/segment - updated_at

kudlohlavec commented:
Please review current state of the branch, all of the requested changes should be already implemented 😉 

Major changes:
- complete removal of laravel scout and its related packages
- incorporation of much more simpler search core that works on top of simple LIKE queries
- no indexing is needed
- added sorting and limiting of returned results

We tried to replicate your dataset in terms of total count of records. Mainly for 70k articles. Our tests showed that results can be retrieved in ~1 second. I consider it to be appropriate to the number of records and leveraging of simple non-indexed LIKE approach. 

Please try responsiveness in your environment if you are ok with it.

kudlohlavec commented:
Sure, I will have a look into it. I am big fan of spatie packages, strange that this one missed me when i was searching for possible alternatives for this feature 😄. I will let you know about further progress on monday 😉

kudlohlavec commented:
Spatie search package looks usable after some initial tests. Indeed, it brought better customization possibilities. 

I need to clarify one more thing - do you want to match results by full or partial wildcard (LIKE %searchTerm% vs LIKE searchTerm%)? The latter provides better leveraging of indexes when using LIKE queries.

rootpd commented:
Latter please.

kudlohlavec commented:
> edit: One of the features I had in mind is returning N latest articles of matched author within the search result (even if they don't match the searched term in article title). I think that Spatie's implementation would allow that.

Yes, it should be possible. But what about presentation of these kind of "linked" results? Current design for returned results is not built for showing relations between separate types of results. Or are you ok if I wil just add more results into the article results section? If so is it ok if it will be already matched results for articles(max 5) + max 5 lattest articles of matched author(s)? I would probably add some check for uniqueness as well if we will decide to go this way. But it can yield many more results if 5 authors are mached and we will be returning 25 lattest articles as well...

rootpd commented:
You've got valid point. As this sounds a bit more complicated, let's skip this feature for the later implementation and revisit this in the future.

kudlohlavec commented:
> You've got valid point. As this sounds a bit more complicated, let's skip this feature for the later implementation and revisit this in the future.

Another possibility came to my mind. We can show related articles right in the returned author row and make them clickable by anchor links. But it still can and probably will look a little bit messy if there are 5 articles with longer title

rootpd commented:
That's not a bad idea. Visually this could be solved by trimming article title to N characters (50?), but I'd probably make this a separate feature/PR as more options will need to be tested there.

kudlohlavec commented:
Agreed

kudlohlavec commented:
Can you please try out the latest commit? Focus mainly on performance, design related tasks are not being implemented yet.

Changelog:
- removed old search core
- added new search core from spatie
- added indexes where needed (see note below)
- improved speed thanks to simpler queries, better execution control and indexing as well

Indexes:
- when Ive been inspecting queries by mysql explain tool, not every query has been using provided indexes so I only added migrations for indexes that made sense
- hard to say what are the reasons behind it but probably some combinations of usage of orderings and limits on queries resulted in dropping of usage of some indexes

kudlohlavec commented:
Design changes already implemented

- added support for 3 additional result attributes coloring to name and title
- i didnt know which colors would you like to use, please refer to package/remp/sass/site-search.scss to test which color palette will be suitable for you
- added progress indicator

rootpd commented:
I've cleaned this up, made some minor changes and merged it to move this forward. We'll be gathering feedback and possible change requests will come as a separate PR for easier tracking and reviewing.

See https://github.com/remp2020/remp/commit/b66c2482cfb7b8a620bdbbe9ed1bddee052d1612 for changes I've made.
